File: app/src/main/java/com/github/alexgreench/observerpattern/MainActivity.java
package com.github.alexgreench.observerpattern;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements RepositoryObserver{

    private Subject dataRepository;
    private TextView textSubscriber;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        dataRepository = DataRepository.getInstance();

        textSubscriber = findViewById(R.id.text_subscriber);

        findViewById(R.id.btn_subscribe).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dataRepository.doSubscribe(MainActivity.this);
            }
        });

        findViewById(R.id.btn_unsubscribe).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dataRepository.doUnsubscribe(MainActivity.this);
                textSubscriber.setText(R.string.you_are_unsubscribed);
            }
        });

    }

    @Override
    public void onDataChanged(String data) {
        textSubscriber.setText(data);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        dataRepository.doUnsubscribe(this);
    }
}

File: app/src/main/java/com/github/alexgreench/observerpattern/MainActivity.kt
package com.github.alexgreench.observerpattern

import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity(), RepositoryObserver {

    private var dataRepository: Subject = DataRepository.instance

    lateinit var textSubscriber: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        textSubscriber = findViewById(R.id.text_subscriber)

        findViewById<View>(R.id.btn_subscribe).setOnClickListener {
            dataRepository.doSubscribe(this@MainActivity)
        }
        findViewById<View>(R.id.btn_unsubscribe).setOnClickListener {
            dataRepository.doUnsubscribe(this@MainActivity)
            textSubscriber.setText(R.string.you_are_unsubscribed)
        }
    }

    override fun onDataChanged(data: String?) {
        textSubscriber.text = data
    }

    override fun onDestroy() {
        super.onDestroy()
        dataRepository.doUnsubscribe(this)
    }
}